device/src/graphics/widgets/enter_share_screen.rs

```diff
@@ -0,0 +1,126 @@
+            touches: vec![],
+            keyboard_rect,
+            input_display_rect,
+        }
+    }
+
+    pub fn draw<D: DrawTarget<Color = Rgb565>>(
+        &mut self,
+        target: &mut D,
+        current_time: Instant<u64, 1, 1_000_000>,
+    ) {
+        self.bech32_keyboard
+            .draw(&mut target.cropped(&self.keyboard_rect));
+        self.backup_input_preview
+            .draw(&mut target.cropped(&self.input_display_rect), current_time);
+
+        self.touches.retain_mut(|touch| {
+            touch.draw(target, current_time);
+            !touch.is_finished()
+        });
+    }
+
+    pub fn handle_touch(
+        &mut self,
+        point: Point,
+        current_time: Instant<u64, 1, 1_000_000>,
+        lift_up: bool,
+    ) {
+        if lift_up {
+            if let Some(active_touch) = self.touches.last_mut() {
+                if let Some(key) = active_touch.let_go(current_time) {
+                    self.backup_input_preview.add_character(key);
+                    if self.backup_input_preview.is_finished() && !self.is_share_valid() {
+                        self.backup_input_preview.set_input_color(COLORS.error);
+                    } else {
+                        self.backup_input_preview.set_input_color(COLORS.primary);
+                    }
+                }
+            }
+        } else {
+            let key_touch = if self.keyboard_rect.contains(point) {
+                let translated_point = point - self.keyboard_rect.top_left;
+                self.bech32_keyboard
+                    .handle_touch(translated_point)
+                    .map(|mut key_touch| {
+                        key_touch.translate(self.keyboard_rect.top_left);
+                        key_touch
+                    })
+            } else {
+                self.backup_input_preview.handle_touch(point)
+            };
+
+            if let Some(key_touch) = key_touch {
+                if let Some(last) = self.touches.last_mut() {
+                    if last.key == key_touch.key {
+                        self.touches.pop();
+                    } else {
+                        last.cancel();
+                    }
+                }
+                self.touches.push(key_touch);
+            }
+        }
+    }
+
+    pub fn is_finished(&self) -> bool {
+        self.backup_input_preview.is_finished()
+    }
+
+    pub fn try_create_share(&self, share_index: u16) -> Result<SecretShare, String> {
+        assert!(self.is_finished(), "must be finished to take share");
+        let characters = self.backup_input_preview.get_input();
+        let backup_string = format!("frost[{}]1{}", share_index, characters.to_lowercase());
+
+        SecretShare::from_bech32_backup(&backup_string).map_err(|_e| backup_string)
+    }
+
+    pub fn is_share_valid(&self) -> bool {
+        if !self.is_finished() {
+            return false;
+        }
+        self.try_create_share(42 /* XXX: doesn't matter to check */)
```

Comment:
I think this does matter for the checksum, if I step through it slowly an error is thrown here after typing in the correct share. One time I think I saw a flash of red characters after entering a valid share, the other call of `try_create_share` with the correct share index means the workflow progresses.

Reply:
We definitely should not be doing this hack but before removing it we owe it to ourselves to figure out why it works at all :rofl:

Reply:
in summary: @nickfarrow was spot on.